numpy/core/src/umath/reduction.c

```diff
@@ -198,38 +193,56 @@ PyUFunc_ReduceWrapper(PyArrayMethod_Context *context,
     /* Loop auxdata (must be freed on error) */
     NpyAuxData *auxdata = NULL;
 
-    /* More than one axis means multiple orders are possible */
-    if (!reorderable && count_axes(PyArray_NDIM(operand), axis_flags) > 1) {
-        PyErr_Format(PyExc_ValueError,
-                     "reduction operation '%s' is not reorderable, "
-                     "so at most one axis may be specified",
-                     funcname);
-        return NULL;
-    }
-    /* Can only use where with an initial ( from identity or argument) */
-    if (wheremask != NULL && identity == Py_None) {
-        PyErr_Format(PyExc_ValueError,
-                     "reduction operation '%s' does not have an identity, "
-                     "so to use a where mask one has to specify 'initial'",
-                     funcname);
-        return NULL;
-    }
-
-
     /* Set up the iterator */
     op[0] = out;
     op[1] = operand;
     op_dtypes[0] = context->descriptors[0];
     op_dtypes[1] = context->descriptors[1];
 
+    /*
+     * Fill in default or identity value and ask if this is reorderable.
+     * Note that if the initial value was provided, we pass `initial_buf=NULL`
+     * to the `get_identity` function to indicate that we only require the
+     * reorderable flag.
+     * If a `result` was passed in, it is possible that the result has a dtype
+     * differing to the operation one.
+     */
+    NPY_ARRAYMETHOD_IDENTITY_FLAGS identity_flags = 0;
+    char *identity_buf = NULL;
+    if (initial == NULL) {
+        /* Always init buffer (only necessary if it holds references) */
+        identity_buf = PyMem_Calloc(1, op_dtypes[0]->elsize);
+        if (identity_buf == NULL) {
+            PyErr_NoMemory();
+            goto fail;
+        }
+    }
+    if (context->method->get_identity(
+            context, identity_buf, &identity_flags) < 0) {
+        PyMem_FREE(identity_buf);  /* get_initial_reduce_value must clean up */
+        goto fail;
+    }
+    /* More than one axis means multiple orders are possible */
+    if (!(identity_flags & NPY_METH_IS_REORDERABLE)
+            && count_axes(PyArray_NDIM(operand), axis_flags) > 1) {
+        PyErr_Format(PyExc_ValueError,
+                "reduction operation '%s' is not reorderable, "
+                "so at most one axis may be specified",
+                funcname);
+        PyMem_FREE(identity_buf);
+        return NULL;
+    }
+
     it_flags = NPY_ITER_BUFFERED |
             NPY_ITER_EXTERNAL_LOOP |
             NPY_ITER_GROWINNER |
-            NPY_ITER_DONT_NEGATE_STRIDES |
             NPY_ITER_ZEROSIZE_OK |
             NPY_ITER_REFS_OK |
             NPY_ITER_DELAY_BUFALLOC |
             NPY_ITER_COPY_IF_OVERLAP;
+    if (!(identity_flags & NPY_METH_IS_REORDERABLE)) {
+        it_flags |= NPY_ITER_DONT_NEGATE_STRIDES;
```

Comment:
I was expecting this change to speed up:
```
arr = np.ones(100000, 3)[::-1]
arr.sum()
```
turns out it doesn't... Something must not be optimized correctly in the iterator construction and now I am wondering if one of my old iterator-cleanup starts might include that fix...

Reply:
Hmm, weird indeed. Still seems nice to add it here.